websocketOutportDriver/sender/webSocketSender.go

```diff
@@ -74,37 +70,30 @@ func NewWebSocketSender(args WebSocketSenderArgs) (*webSocketSender, error) {
 	return ws, nil
 }
 
+// AddClient will add the client to internal maps and will also start
 func (w *webSocketSender) AddClient(wss WSConn, remoteAddr string) {
 	client := &webSocketClient{
 		conn:       wss,
 		remoteAddr: remoteAddr,
 	}
-	w.mutClients.Lock()
-	w.clients[remoteAddr] = client
-	w.mutClients.Unlock()
 
-	w.mutAcknowledges.Lock()
-	w.acknowledges[remoteAddr] = make(map[uint64]struct{})
-	w.mutAcknowledges.Unlock()
+	w.clientsHolder.AddClient(client)
 
-	go w.handleReceiveAck(client)
-}
+	w.acknowledges.AddEntryForClient(remoteAddr)
 
-func (w *webSocketSender) getClients() map[string]*webSocketClient {
-	w.mutClients.RLock()
-	defer w.mutClients.RUnlock()
+	if !w.withAcknowledge {
+		return
+	}
 
-	return w.clients
+	go w.handleReceiveAck(client)
 }
 
 func (w *webSocketSender) handleReceiveAck(client *webSocketClient) {
 	for {
 		mType, message, err := client.conn.ReadMessage()
 		if err != nil {
 			w.log.Error("cannot read message", "remote addr", client.remoteAddr, "error", err)
-			w.mutClients.Lock()
-			delete(w.clients, client.remoteAddr)
-			w.mutClients.Unlock()
+			w.clientsHolder.Remove(client.remoteAddr)
```

Comment:
Shouldn't you also remove the entry from `acknowledges`? 
If so, you should do it under a mutex so it is done atomically.

Reply:
done. thanks

Reply:
>If so, you should do it under a mutex so it is done atomically.

This is not solved. You would probably have to:
```go
w.mutex.Lock()
w.clientsHolder.Remove(client.remoteAddr)
w.acknowledges.RemoveEntryForAddress(client.remoteAddr)
w.Mutex.Unlock()
```

Also, I think that on `clientsHolder.Remove`, before delete, you should also: `	wch.clients[remoteAddr].conn.Close()`

Reply:
no need for a mutex as the clients and the acknowledges components are somehow independent and they all have mutexes inside. Renamed the `Remove` function to `CloseAndRemove` that does what it's name says